File: bst/levelorder/treedepth/DepthBTRecursive.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class DepthBTRecursive:
    def maxDepth(self, root: TreeNode) -> int:
        # Base case: if the node is null, the depth is 0
        if not root:
            return 0
        # Recursively find the depth of left and right subtrees and return the max depth + 1
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

# Example usage:
if __name__ == "__main__":
    solution = DepthBTRecursive()

    # Example 1
    root1 = TreeNode(3)
    root1.left = TreeNode(9)
    root1.right = TreeNode(20)
    root1.right.left = TreeNode(15)
    root1.right.right = TreeNode(7)
    print(solution.maxDepth(root1))  # Output: 3

    # Example 2
    root2 = TreeNode(1)
    root2.right = TreeNode(2)
    print(solution.maxDepth(root2))  # Output: 2

File: bst/levelorder/treedepth/DepthBTRecursive.java
package levelorder.treedepth;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class DepthBTRecursive {
    public int maxDepth(TreeNode root) {
        // Base case: if the node is null, the depth is 0
        if (root == null) {
            return 0;
        }
        // Recursively find the depth of the left and right subtrees
        int leftDepth = maxDepth(root.left);
        int rightDepth = maxDepth(root.right);

        // The depth of the current node is 1 + the max depth of the left and right subtrees
        return Math.max(leftDepth, rightDepth) + 1;
    }

    public static void main(String[] args) {
        DepthBTRecursive solution = new DepthBTRecursive();

        // Example 1
        TreeNode root1 = new TreeNode(3);
        root1.left = new TreeNode(9);
        root1.right = new TreeNode(20);
        root1.right.left = new TreeNode(15);
        root1.right.right = new TreeNode(7);
        System.out.println(solution.maxDepth(root1));  // Output: 3

        // Example 2
        TreeNode root2 = new TreeNode(1);
        root2.right = new TreeNode(2);
        System.out.println(solution.maxDepth(root2));  // Output: 2
    }
}
